File: Python/lowbass_ai_painting_cog.py
"""
lowbass_ai_painting_cog.py
-------------------------
Lowbass' AI Painting Cog - a **stand-alone** TwitchIO extension that adds an AI-powered "painting" feature to any existing bot.

Features
~~~~~~~~
* **Admin-only !paint command** - broadcasters & mods can generate paintings on demand.

Environment variables
~~~~~~~~~~~~~~~~~~~~~
OPENAI_API_KEY       # required - key for OpenAI Images API
IMGUR_CLIENT_ID      # required - Imgur API client ID
IMGUR_ALBUM_ID       # optional - album to contain generated paintings
BOT_BROADCASTER      # required - Twitch username of the broadcaster (for permission checks)
"""

import os
import openai
import pyimgur
import requests
from twitchio.ext import commands

class PaintingCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        print("[PaintingCog] Initializing cog...")

        # Load OpenAI API key from environment variables
        openai.api_key = os.getenv("OPENAI_API_KEY")
        print("[PaintingCog] OpenAI API key loaded.")

        # Initialize Imgur client with client ID from environment variables
        self.imgur_client = pyimgur.Imgur(os.getenv("IMGUR_CLIENT_ID"))
        print("[PaintingCog] Imgur client initialized.")

        # Optionally use a specific Imgur album to store uploaded images
        self.imgur_album_id = os.getenv("IMGUR_ALBUM_ID")

        # Store broadcaster's Twitch username for permission checks
        self.broadcaster = os.getenv("BOT_BROADCASTER")
        print(f"[PaintingCog] Broadcaster set to: {self.broadcaster}")

    @commands.command(name="paint")
    async def paint_command(self, ctx: commands.Context):
        print(f"[PaintingCog] Received !paint command from: {ctx.author.name}")

        # Only allow the broadcaster to use the !paint command
        if ctx.author.name.lower() != self.broadcaster.lower():
            print("[PaintingCog] Permission denied for user.")
            await ctx.send(f"Sorry {ctx.author.name}, only the broadcaster can use this command.")
            return

        # Parse the command for user and prompt
        content = ctx.message.content.strip()
        split_content = content.split(" ", 2)
        if len(split_content) < 3:
            print("[PaintingCog] Invalid command format.")
            await ctx.send("Usage: !paint <user> <prompt>")
            return

        user = split_content[1]
        prompt = split_content[2]
        print(f"[PaintingCog] Generating painting for user: {user}, prompt: {prompt}")
        await ctx.send(f"🎨 Generating painting for {user}: '{prompt}'…")

        try:
            # Generate and upload the painting, then get the image URL
            url = self.generate_and_upload_painting(prompt)
            await ctx.send(f"{user}, here is your painting: {url}")
        except Exception as e:
            print(f"[PaintingCog] Error during painting generation: {e}")
            await ctx.send(f"Sorry, an error occurred: {e}")

    def generate_and_upload_painting(self, prompt):
        print(f"[PaintingCog] Generating painting for prompt: {prompt}")

        # Use OpenAI's image generation API to create an image based on the prompt
        response = openai.Image.create(
            prompt=prompt,
            n=1,
            size="1024x1024"
        )

        # Extract the image URL from the API response
        image_url = response['data'][0]['url']
        print(f"[PaintingCog] Downloading generated image from {image_url}")

        # Download the image data from the generated URL
        image_data = requests.get(image_url).content

        # Save the image temporarily to disk
        temp_file_path = "/tmp/painting.png"
        with open(temp_file_path, "wb") as f:
            f.write(image_data)

        print("[PaintingCog] Uploading to Imgur…")

        # Upload the image to Imgur, optionally associating it with a specific album
        uploaded_image = self.imgur_client.upload_image(
            temp_file_path,
            album=self.imgur_album_id,
            title="AI Painting"
        )

        print(f"[PaintingCog] Upload complete: {uploaded_image.link}")
        # Return the Imgur URL of the uploaded image
        return uploaded_image.link

File: JavaScript/lowbass_ai_painting_cog.js
// lowbass_ai_painting_cog.js
// ---------------------------
// Lowbass' AI Painting Cog – a standalone module for integrating an AI-powered painting command into a Twitch bot.

const tmi = require('tmi.js'); // Import the tmi.js library for Twitch chat interaction
const axios = require('axios'); // Import axios for making HTTP requests
const fs = require('fs'); // Import the file system module for file operations
const path = require('path'); // Import the path module for handling file paths
require('dotenv').config(); // Load environment variables from a .env file

// Load environment variables for API keys and other configurations
const IMGUR_CLIENT_ID = process.env.IMGUR_CLIENT_ID; // Imgur API client ID
const IMGUR_ALBUM_ID = process.env.IMGUR_ALBUM_ID; // Optional Imgur album ID
const OPENAI_API_KEY = process.env.OPENAI_API_KEY; // OpenAI API key
const BROADCASTER = process.env.BOT_BROADCASTER; // Twitch username of the broadcaster

// Function to initialize the bot with the painting feature
function initializePaintingFeature(client) {
  console.log("[PaintingFeature] Initializing...");

  // Listen for messages in the Twitch chat
  client.on('message', async (channel, tags, message, self) => {
    if (self) return; // Ignore messages from the bot itself

    // Only respond to messages that start with the !paint command
    if (!message.startsWith('!paint')) return;

    // Get the username of the message sender
    const username = tags['display-name'] || tags.username;
    console.log(`[PaintingFeature] Command received from: ${username}`);

    // Check if the user is the broadcaster; if not, deny permission
    if (username.toLowerCase() !== BROADCASTER.toLowerCase()) {
      console.log("[PaintingFeature] Permission denied.");
      return;
    }

    // Split the message into parts and check if it has enough arguments
    const parts = message.trim().split(' ').slice(1);
    if (parts.length < 2) {
      client.say(channel, "Usage: !paint <user> <prompt>"); // Inform the user of the correct usage
      return;
    }

    // Extract the target user and the painting prompt from the message
    const user = parts[0];
    const prompt = parts.slice(1).join(' ');
    console.log(`[PaintingFeature] Generating painting for ${user}, prompt: ${prompt}`);

    // Notify the channel that the painting is being generated
    client.say(channel, `🎨 Generating painting for ${user}: '${prompt}'...`);

    try {
      // Generate and upload the painting, then send the image URL to the channel
      const imageUrl = await generateAndUploadPainting(prompt);
      client.say(channel, `${user}, here is your painting: ${imageUrl}`);
    } catch (error) {
      // Log and notify the channel of any errors that occur
      console.error("[PaintingFeature] Error generating or uploading image:", error);
      client.say(channel, `Sorry, an error occurred: ${error.message}`);
    }
  });
}

// Function to generate and upload a painting based on a prompt
async function generateAndUploadPainting(prompt) {
  console.log(`[PaintingFeature] Sending prompt to OpenAI: ${prompt}`);

  // Request an image from OpenAI using the provided prompt
  const openaiResp = await axios.post(
    'https://api.openai.com/v1/images/generations',
    {
      prompt, // The prompt for the image generation
      n: 1, // Number of images to generate
      size: "1024x1024" // Size of the generated image
    },
    {
      headers: {
        'Authorization': `Bearer ${OPENAI_API_KEY}`, // Authorization header with OpenAI API key
        'Content-Type': 'application/json' // Content type for the request
      }
    }
  );

  // Extract the image URL from the OpenAI response
  const imageUrl = openaiResp.data.data[0].url;
  console.log(`[PaintingFeature] Image received from OpenAI: ${imageUrl}`);

  // Download the image locally to a temporary file
  const imageResp = await axios.get(imageUrl, { responseType: 'arraybuffer' });
  const filePath = path.join(__dirname, 'temp_painting.png'); // Path for the temporary file
  fs.writeFileSync(filePath, imageResp.data); // Save the image data to the file
  console.log("[PaintingFeature] Image saved locally.");

  // Upload the image to Imgur
  const imgurResp = await axios.post(
    'https://api.imgur.com/3/image',
    {
      image: fs.readFileSync(filePath, { encoding: 'base64' }), // Read and encode the image file
      type: 'base64', // Specify the image type
      album: IMGUR_ALBUM_ID, // Optional album ID for Imgur
      title: 'AI Painting' // Title for the uploaded image
    },
    {
      headers: {
        Authorization: `Client-ID ${IMGUR_CLIENT_ID}` // Authorization header with Imgur client ID
      }
    }
  );

  // Extract the uploaded image URL from the Imgur response
  const uploadedUrl = imgurResp.data.data.link;
  console.log(`[PaintingFeature] Image uploaded to Imgur: ${uploadedUrl}`);

  return uploadedUrl; // Return the URL of the uploaded image
}

module.exports = initializePaintingFeature; // Export the initialize function for use in other modules